compiler+runtime/test/cpp/jank/read/lex.cpp

```diff
@@ -467,6 +487,104 @@ namespace jank::read::lex
+      {
+        processor p{ "0x34 0Xab 0x12ab 123 0Xef43 -0x1a" };
+        native_vector<result<token, error>> tokens(p.begin(), p.end());
+        CHECK(tokens
+              == make_tokens({
+                {  0, 4, token_kind::integer,    52ll },
+                {  5, 4, token_kind::integer,   171ll },
+                { 10, 6, token_kind::integer,  4779ll },
+                { 17, 3, token_kind::integer,   123ll },
+                { 21, 6, token_kind::integer, 61251ll },
+                { 28, 5, token_kind::integer,   -26ll }
+        }));
+      }
+
+      SUBCASE("Invalid hex numbers")
+      {
+        processor p{ "0xg 0x-2 0x8.4 0x3e-5 0x 1x" };
+        native_vector<result<token, error>> tokens(p.begin(), p.end());
+        CHECK(tokens
+              == make_results({
+                error{  0,  3,          "invalid number: char g are invalid for radix 16" },
+                error{  4,  8,          "invalid number: char - are invalid for radix 16" },
+                error{  9, 14,          "invalid number: char . are invalid for radix 16" },
+                error{ 15, 21,          "invalid number: char - are invalid for radix 16" },
+                error{ 22, 24, "unexpected end of radix 16 number, expecting more digits" },
+                error{ 25, 27,          "invalid number: char x are invalid for radix 10" }
+        }));
+      }
+
+      SUBCASE("Valid arbitrary radix")
+      {
+        processor p{ "2r11 36rz 8R71 19rghi -4r32 16r3e 16r3e4 -32r3e4" };
+        native_vector<result<token, error>> tokens(p.begin(), p.end());
+        CHECK(tokens
+              == make_results({
+                token{  0, 4, token_kind::integer,     3ll },
+                token{  5, 4, token_kind::integer,    35ll },
+                token{ 10, 4, token_kind::integer,    57ll },
+                token{ 15, 6, token_kind::integer,  6117ll },
+                token{ 22, 5, token_kind::integer,   -14ll },
+                token{ 28, 5, token_kind::integer,    62ll },
+                token{ 34, 6, token_kind::integer,   996ll },
+                token{ 41, 7, token_kind::integer, -3524ll },
+        }));
+      }
+
+      SUBCASE("Invalid arbitrary radix")
+      {
+        processor p{ "2r3 35rz 8re71 19r-ghi 2r 16r" };
+        native_vector<result<token, error>> tokens(p.begin(), p.end());
+        CHECK(tokens
+              == make_results({
+                error{  0,  3,        "invalid number: char 3 are invalid for radix 2" },
+                error{  4,  8,       "invalid number: char z are invalid for radix 35" },
+                error{  9, 14,        "invalid number: char e are invalid for radix 8" },
+                error{ 15, 22,       "invalid number: char - are invalid for radix 19" },
+                error{ 23, 25, "unexpected end of radix number, expecting more digits" },
+                error{ 26, 29, "unexpected end of radix number, expecting more digits" }
+        }));
+      }
```

Comment:
I really like how you've captured both positive and negative tests here. Let's dig deeper into potential negative cases, since I think there are more. Off the top of my head, to get the ball rolling:

* `0r0` (is radix 0 supported?) -- In Clojure, no
* `1r1` (is radix 1 supported?) -- In Clojure, no
* `2048r0` (how high can radix actually go?) -- In Clojure, `36r0` is the highest
* `2.0r1` (what if radix is a real?)
* `2rr1` (what if `r` is specified multiple times?)
* `2r1r` (what if `r` is specified at the end of a valid number?)
* `2r.0` (what if a decimal is specified after `r`?)

Reply:
Please also give the same treatment to octal and hex. I can help with some test cases if you need more suggestions.

Reply:
updated with more unit tests.